library/ccm.c

```diff
@@ -291,60 +341,104 @@ static int ccm_auth_crypt( mbedtls_ccm_context *ctx, int mode, size_t length,
     src = input;
     dst = output;
-
+    if( len_left == 0 )
+    {
+        fc3 = 0;
+        fc4 = 0;
+        fc5 = 0;
+    }
     while( len_left > 0 )
     {
         size_t use_len = len_left > 16 ? 16 : len_left;
 
         if( mode == CCM_ENCRYPT )
         {
-            mbedtls_platform_memset( b, 0, 16 );
-            mbedtls_platform_memcpy( b, src, use_len );
-            UPDATE_CBC_MAC;
+            mbedtls_platform_random_delay();
+            if( mode_dup == CCM_ENCRYPT )
+            {
+                mbedtls_platform_memset( b, 0, 16 );
+                mbedtls_platform_memcpy( b, src, use_len );
+                UPDATE_CBC_MAC;
+                CHECK_FLOW_CTRL( 16, fc3, flow_ctrl_local );
+                fc5 = 0;
+            }
+            else
+            {
+                return( MBEDTLS_ERR_PLATFORM_FAULT_DETECTED );
+            }
         }
 
         CTR_CRYPT( dst, src, use_len );
+        CHECK_FLOW_CTRL( use_len, fc4, flow_ctrl_local );
 
         if( mode == CCM_DECRYPT )
         {
-            mbedtls_platform_memset( b, 0, 16 );
-            mbedtls_platform_memcpy( b, dst, use_len );
-            UPDATE_CBC_MAC;
+            mbedtls_platform_random_delay();
+            if( mode_dup == CCM_DECRYPT )
+            {
+                mbedtls_platform_memset( b, 0, 16 );
+                mbedtls_platform_memcpy( b, dst, use_len );
+                UPDATE_CBC_MAC;
+                CHECK_FLOW_CTRL( 16, fc5, flow_ctrl_local );
+                fc3 = 0;
+            }
+            else
+            {
+                return( MBEDTLS_ERR_PLATFORM_FAULT_DETECTED );
+            }
         }
 
         dst += use_len;
         src += use_len;
         len_left -= use_len;
-
         /*
          * Increment counter.
          * No need to check for overflow thanks to the length check above.
          */
         for( i = 0; i < q; i++ )
             if( ++ctr[15-i] != 0 )
                 break;
+        for( i = 0; i < q; i++ )
+            if( ++ctr_dup[15-i] != 0 )
+                break;
+
+        if( mbedtls_platform_memcmp( ctr_dup, ctr, 16 ) != 0 )
+            return( MBEDTLS_ERR_PLATFORM_FAULT_DETECTED );
     }
 
+    if( len_left > 0 )
+    {
+        return( MBEDTLS_ERR_PLATFORM_FAULT_DETECTED );
+    }
     /*
      * Authentication: reset counter and crypt/mask internal tag
      */
+    flow_ctrl_local = 0;
     for( i = 0; i < q; i++ )
+    {
+        flow_ctrl_local++;
         ctr[15-i] = 0;
+        ctr_dup[15-i] = 0;
+    }
+    CHECK_FLOW_CTRL( q, fc6, flow_ctrl_local );
 
     CTR_CRYPT( y, y, 16 );
+    CHECK_FLOW_CTRL( 16, fc7, flow_ctrl_local );
     mbedtls_platform_memcpy( tag, y, tag_len );
 
-    if( length_dup != length || iv_dup != iv || iv_len_dup != iv_len ||
-        add_dup != add || add_len_dup != add_len || input_dup != input ||
-        output_dup != output || tag_dup != tag || tag_len_dup != tag_len)
-    {
+    if( mbedtls_platform_memcmp( ctr_dup, ctr, 16 ) != 0 )
```

Comment:
can we compare it only once in this line and omit comparison in line 405?

Reply:
The earlier comparison is done in a loop, after every step of at most 16 bytes.